src/Interpreters/Session.h

```diff
@@ -0,0 +1,82 @@
+#pragma once
+
+#include <common/types.h>
+#include <Interpreters/Context_fwd.h>
+#include <Interpreters/ClientInfo.h>
+
+#include <chrono>
+#include <memory>
+#include <optional>
+
+namespace Poco::Net { class SocketAddress; }
+
+namespace DB
+{
+class SessionLog;
+class Credentials;
+struct NamedSession;
+class ContextAccess;
+struct Settings;
+class Authentication;
+
+/** Represents user-session from the server perspective,
+ *  basically it is just a smaller subset of Context API, simplifies Context management.
+ *
+ * Holds session context, facilitates acquisition of NamedSession and proper creation of query contexts.
+ * Adds log in, log out and login failure events to the SessionLog.
+ */
+class Session
+{
+public:
+//    static Session makeSessionFromCopyOfContext(const ContextPtr & _context_to_copy);
+    Session(const ContextPtr & context_to_copy, ClientInfo::Interface interface, std::optional<String> default_format = std::nullopt);
+    ~Session();
+
+    Session(const Session &) = delete;
+    Session& operator=(const Session &) = delete;
+
+    Session(Session &&);
+    Session& operator=(Session &&);
+
+    Authentication getUserAuthentication(const String & user_name) const;
+    void setUser(const Credentials & credentials, const Poco::Net::SocketAddress & address);
+    void setUser(const String & name, const String & password, const Poco::Net::SocketAddress & address);
+
+    // Manually handle login failure event in complex login scenarious.
+    void onLogInFailure(const String & user_name, const String & failure_reason);
+
+    /** Propmotes current session to a named session.
+     *
+     *  that is: re-uses or creates NamedSession and then piggybacks on it's context,
+     *  retaining ClientInfo of current session_context.
+     *  Acquired named_session is then released in the destructor.
+     */
+    void promoteToNamedSession(const String & session_id, std::chrono::steady_clock::duration timeout, bool session_check);
+    /// Early release a NamedSession.
+    void releaseNamedSession();
+
+    ContextMutablePtr makeQueryContext(const String & query_id) const;
+
+    ContextPtr sessionContext() const;
+    ContextMutablePtr mutableSessionContext();
+
+    ClientInfo & getClientInfo();
+    const ClientInfo & getClientInfo() const;
+
+    const Settings & getSettings() const;
+
+    void setQuotaKey(const String & quota_key);
+
+    String getCurrentDatabase() const;
+    void setCurrentDatabase(const String & name);
+
+protected:
+    std::shared_ptr<SessionLog> getSessionLog();
+
+private:
+    ContextMutablePtr session_context;
+    std::shared_ptr<const ContextAccess> access;
+    std::shared_ptr<NamedSession> named_session;
+};
```

Comment:
So now we have two separate classes: `Session` and `NamedSession` which look completely different. It doesn't look good to me. Can we somehow combine them? For example we could add all the fields from `NamedSession` into `Session` and make the class `NamedSessions` to keep `Session`s instead of `NamedSession`s.

Reply:
Removed `NamedSession` from Context.cpp, and now we have a `NamedSessionData` in the Session.cpp.